MNT/ENH: Add bucket replication configuration to the CDK

greglucas opened the pull request:
Right now we have this specified as manual steps to go in and update it in the console. We can create the replication rule with L1 constructs, so not as elegant as the L2 methods, but still doable in code rather than manual intervention.

The backup bucket arn needs to exist for this deployment to work.

greglucas committed:
MNT/ENH: Add bucket replication configuration to the CDK

Right now we have this specified as manual steps to go in and update it.
We can create the replication rule with L1 constructs, so not as elegant
as the L2 methods, but still doable in code rather than manual intervention.

maxinelasp reviewed (approved):
Looks good, nice update! What if you deploy this and the backup bucket arn doesn't exist?

maxinelasp commented:
Does this create a circular dependency with the backup role, actually? The backup role needs to exist before the table does, but I think it's for the main bucket, so the main bucket needs to be created first. Then the backup bucket uses the backup role to point to the main bucket, but then this piece points to the backup bucket so it can't actually work. I think instead we need to have this try to create, if the backup bucket doesn't exist then continue on anyway and skip that replication step, and then the basic steps would be: 
1. deploy this bucket, have the replication bit fail gracefully
2. Deploy backup bucket with the correct permissions
3. re-deploy this bucket, now with replication succeeding

Or, maybe put replication into a different stack entirely

maxinelasp commented:
Alternatively, Sean mentioned just having the backup bucket in the same account but in a different region - that would also solve all these problems and a bunch of other problems

greglucas commented:
If the bucket doesn't exist, this will not deploy. There is an error that the backup bucket arn must exist and it rolls back gracefully to the previous state.

Good question about the circular reference. I tested this in the same account with a different test bucket arn that I created manually and that did work. Can we deploy the backup stack first? I think we could actually update the backup stack to infer the BackupRole ARN similar to how we are doing it for the BackupBucket ARN, which would remove circular references?

https://github.com/IMAP-Science-Operations-Center/sds-data-manager/blob/2ac3ca4f1f82986eb75d2a83895efae93cde3304/sds_data_manager/constructs/backup_bucket_construct.py#L52-L55
Change that to:
```python
role_arn = f"arn:aws:iam::{source_account}:role/BackupRole"
```

Will the backup stack deploy with that change, or does that role need to exist first so we are in a chicken and egg situation?

maxinelasp commented:
Actually, looking at the current BackupRole, you're right that something changed (presumably in the stack collapsing refactor) that means we can assume the BackupRole is just what you said. Before, it always added some random CDK numbers to it, but now it looks like the role just is named BackupRole. So that solves the problem, I think, as long as the backup is always deployed. 

I do think we should have the option to successfully deploy (maybe with lots of warnings? idk) without the backup bucket because we don't want that in dev.

greglucas committed:
MNT: Make backup addition dependent on account environment

This adds a flag testing whether we are deploying into dev, prod, ...
and only adds the replication policy to production.

greglucas commented:
OK, I added in a check for whether we are in the "sit-4" or "prod" account_names and only creating the backup roles in those situations, so avoiding creating any of the backup stuff at all in the dev account.

I updated the documentation to reflect the current state of things.

I'm still a bit unclear whether the backup stack can deploy without the production stack/bucket or not. Can you verify whether that role needs to be there or not? If not, then I think we can say that the backup stack must be deployed first and then the production stack can be deployed and all will be good.

maxinelasp commented:
Just tested it, I ended up in kind of a weird state but I _think_ everything deployed properly...? I can test more after the SIT.

maxinelasp commented:
So I think we can deploy backup first and then do the main one

tech3371 commented on docs/source/cdk/s3-replication.rst:
```diff
@@ -1,31 +1,21 @@
+To restore files from the backup bucket into the main account bucket, we are going to
+[assume the role](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-role.html#cli-role-prereqs)
+created by SdsDataManager stack in the source account. Here, "source account" refers to
+the account with SdsDataManager deployed into it (i.e. dev or prod).
```
```suggestion
created by SDCStack stack in the source account. Here, "source account" refers to
the account with SDCStack deployed into it (i.e. dev or prod).
```

tech3371 commented on sds_data_manager/constructs/backup_bucket_construct.py:
```diff
@@ -14,15 +14,7 @@ class BackupBucket(Construct):
     should be the AWS account for the source bucket.
 
     For replication to work, you also need to deploy SdsDataManager and create
```
Can you replace `SdsDataManager` with `SDCStack`? It must be remains from a while ago.
```suggestion
    For replication to work, you also need to deploy SDCStack and create
```

tech3371 reviewed (approved):
couple minor comments. Otherwise, looks good to me on high-level.

greglucas committed:
Merge branch 'dev' into bucket-replication

greglucas committed:
Update docs/source/cdk/s3-replication.rst

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

greglucas committed:
Update sds_data_manager/constructs/backup_bucket_construct.py

Co-authored-by: Tenzin Choedon <36522642+tech3371@users.noreply.github.com>

greglucas commented:
This is out of date, I'm going to close it and we can add it back in later if we want. I won't delete the branch yet.
